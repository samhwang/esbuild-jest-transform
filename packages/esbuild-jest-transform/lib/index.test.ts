import * as fs from 'node:fs';
import * as path from 'node:path';
import { TransformOptions } from '@jest/transform';
import { defaults } from 'jest-config';
import transformer, { Options } from './index';

function getTransformConfig(sourcePath: string, options?: Options) {
  const Transformer = transformer.createTransformer({
    format: 'cjs',
    sourcemap: true,
    ...options,
  });

  const config: TransformOptions = {
    config: {
      ...defaults,
      cwd: path.resolve(),
    },
  } as any;

  return { Transformer, config };
}

function getExampleFile(filePath: string) {
  const examplePath = path.resolve(__dirname, '..', '..', '..', 'examples');

  return path.resolve(examplePath, filePath);
}

describe('Transformer tests', () => {
  const nodeFile = getExampleFile('node/index.test.ts');

  describe('Process Sync', () => {
    function processSync(sourcePath: string, options?: Options) {
      const content = fs.readFileSync(sourcePath, 'utf-8');

      const { Transformer, config } = getTransformConfig(sourcePath, options);

      return Transformer.process!(content, sourcePath, config) as {
        code: string;
        map: string;
      };
    }

    it('Should have sourcemap inline as default', () => {
      const output = processSync(nodeFile, { sourcemap: undefined });
      expect(output.map).toEqual('');
      expect(output.code).toContain(
        '//# sourceMappingURL=data:application/json;base64,'
      );
    });

    it('Should correctly get sourcemap as inline if provided', () => {
      const output = processSync(nodeFile, { sourcemap: 'inline' });
      expect(output.map).toEqual('');
      expect(output.code).toContain(
        '//# sourceMappingURL=data:application/json;base64,'
      );
    });

    it('Should get sourcemap as object if true is passed', () => {
      const output = processSync(nodeFile, { sourcemap: true });
      expect(output.map).toHaveProperty('version');
      expect(output.map).toHaveProperty('sources');
      expect(output.map).toHaveProperty('mappings');
      expect(output.map).toHaveProperty('names');
    });

    it('Should match transformed snapshot for node file', () => {
      const output = processSync(nodeFile);
      expect(output.code).toMatchSnapshot();
    });

    it('Should match transformed snapshot for react classic file', () => {
      const reactClassicFuncComp = getExampleFile(
        'react-classic/src/FuncComp.test.tsx'
      );

      const reactClassicClassComp = getExampleFile(
        'react-classic/src/FuncComp.test.tsx'
      );

      const output1 = processSync(reactClassicFuncComp);
      expect(output1.code).toMatchSnapshot();

      const output2 = processSync(reactClassicClassComp);
      expect(output2.code).toMatchSnapshot();
    });
  });

  describe('Process async', () => {
    function processAsync(sourcePath: string, options?: Options) {
      const content = fs.readFileSync(sourcePath, 'utf-8');

      const { Transformer, config } = getTransformConfig(sourcePath, options);

      return Transformer.processAsync!(content, sourcePath, config) as Promise<{
        code: string;
        map: string;
      }>;
    }

    it('Should match transformed snapshot', async () => {
      const output = await processAsync(nodeFile, { sourcemap: true });
      expect(output.code).toMatchSnapshot();
    });
  });
});
