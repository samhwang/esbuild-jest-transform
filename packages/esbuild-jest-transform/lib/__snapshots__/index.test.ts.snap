// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transformer tests Process Sync Should match transformed snapshot 1`] = `
"{
  \\"version\\": 3,
  \\"sources\\": [\\"/Users/samhwang/projects/personal/esbuild-jest/esbuild-jest-transform/examples/node-ts/index.test.ts\\"],
  \\"mappings\\": \\"AAAA,mBAA0B;AAE1B,GAAG,8BAA8B,MAAM;AACrC,QAAM,SAAS,4BAAU,GAAG,CAAC;AAC7B,SAAO,MAAM,EAAE,QAAQ,CAAC;AAC1B,CAAC;\\",
  \\"names\\": []
}
"
`;

exports[`Transformer tests Process Sync Should match transformed snapshot 2`] = `
"var import_index = require(\\"./index\\");
it(\\"Should calculate 1 + 1 = 2\\", () => {
  const result = (0, import_index.addNumber)(1, 1);
  expect(result).toEqual(2);
});
"
`;

exports[`Transformer tests Process async Should match transformed snapshot 1`] = `
"{
  \\"version\\": 3,
  \\"sources\\": [\\"/Users/samhwang/projects/personal/esbuild-jest/esbuild-jest-transform/examples/node-ts/index.test.ts\\"],
  \\"mappings\\": \\"AAAA;AAEA,GAAG,8BAA8B,MAAM;AACrC,QAAM,SAAS,UAAU,GAAG,CAAC;AAC7B,SAAO,MAAM,EAAE,QAAQ,CAAC;AAC1B,CAAC;\\",
  \\"names\\": []
}
"
`;

exports[`Transformer tests Process async Should match transformed snapshot 2`] = `
"import { addNumber } from \\"./index\\";
it(\\"Should calculate 1 + 1 = 2\\", () => {
  const result = addNumber(1, 1);
  expect(result).toEqual(2);
});
"
`;
